sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -714,6 +714,17 @@ object TypeCoercion {
         // try to find the first one we can implicitly cast.
         case (_, TypeCollection(types)) => types.flatMap(implicitCast(e, _)).headOption.orNull
 
+        case (ArrayType(_, nullable), ArrayType(internalType: DataType, expectedNullable))
+          if (expectedNullable || nullable == expectedNullable) =>
```

Comment:
Thank you for your detailed explanation and yeap that makes sense. I see it is up to what `None` means from the function. I was just worried of changing the existing behaviour and wanted to note just in case. I agree with either way if it sounds consistent:).